File: src/test/java/com/github/mvysny/vaadinsimplesecurity/util/PasswordHashTest.java
package com.github.mvysny.vaadinsimplesecurity.util;

import org.junit.jupiter.api.Test;

import static com.github.mvysny.vaadinsimplesecurity.util.PasswordHash.createHash;
import static com.github.mvysny.vaadinsimplesecurity.util.PasswordHash.validatePassword;
import static org.junit.jupiter.api.Assertions.*;

class PasswordHashTest {
    @Test
    public void testMatchingPassword() {
        assertTrue(validatePassword("foo", createHash("foo")));
    }

    @Test
    public void testNonMatchingPassword() {
        assertFalse(validatePassword("bar", createHash("foo")));
    }

    @Test
    public void testDifferentSalt() {
        final String hash1 = createHash("foo".toCharArray(), "salt1".getBytes());
        String hash2 = createHash("foo".toCharArray(), "salt2".getBytes());
        assertNotEquals(hash1, hash2);
    }

    @Test
    public void testDifferentPassword() {
        String hash1 = createHash("foo".toCharArray(), "salt1".getBytes());
        String hash2 = createHash("bar".toCharArray(), "salt1".getBytes());
        assertNotEquals(hash1, hash2);
    }

    @Test
    public void originalPasswordHashTest() {
        // Print out 10 hashes
        for (int i = 0; i < 10; i++)
            System.out.println(createHash("p\r\nassw0Rd!"));

        // Test password validation
        for (int i = 0; i < 10; i++) {
            String password = "" + i;
            String hash = createHash(password);
            String secondHash = createHash(password);
            assertNotEquals(hash, secondHash);
            String wrongPassword = "" + (i + 1);
            assertFalse(validatePassword(wrongPassword, hash), "FAILURE: WRONG PASSWORD ACCEPTED!");
            assertTrue(validatePassword(password, hash), "FAILURE: GOOD PASSWORD NOT ACCEPTED!");
        }
    }
}

File: vaadin-simple-security/src/test/java/com/github/mvysny/vaadinsimplesecurity/util/PasswordHashTest.kt
package com.github.mvysny.vaadinsimplesecurity.util

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import kotlin.test.expect

class PasswordHashTest {
    @Test fun matchingPassword() {
        expect(true) {
            PasswordHash.validatePassword("foo", PasswordHash.createHash("foo"))
        }
    }

    @Test fun `non-matching password`() {
        expect(false) {
            PasswordHash.validatePassword("bar", PasswordHash.createHash("foo"))
        }
    }

    @Test fun differentSalt() {
        val hash1 = PasswordHash.createHash("foo".toCharArray(), "salt1".toByteArray())
        val hash2 = PasswordHash.createHash("foo".toCharArray(), "salt2".toByteArray())
        expect(false, "$hash1, $hash2") { hash1 == hash2 }
    }

    @Test fun differentPassword() {
        val hash1 = PasswordHash.createHash("foo".toCharArray(), "salt1".toByteArray())
        val hash2 = PasswordHash.createHash("bar".toCharArray(), "salt1".toByteArray())
        expect(false, "$hash1, $hash2") { hash1 == hash2 }
    }

    @Test fun originalPasswordHashTest() {
        // Print out 10 hashes
        for (i in 0..9) println(PasswordHash.createHash("p\r\nassw0Rd!"))

        // Test password validation
        for (i in 0..9) {
            val password = "" + i
            val hash = PasswordHash.createHash(password)
            val secondHash = PasswordHash.createHash(password)
            Assertions.assertNotEquals(hash, secondHash)
            val wrongPassword = "" + (i + 1)
            expect(false) { PasswordHash.validatePassword(wrongPassword, hash) }
            expect(true) { PasswordHash.validatePassword(password, hash) }
        }
    }
}
